### ISSUE TITLE
Training Time for 1x 4090 GPU

### ISSUE THREAD

**Issue by akjindal53244** [karpathy/nanoGPT#179]
Hi there, I am running on 1x 4090 GPU and using pytorch 2.0 nighty build. Running `python train.py` as suggested with default parameters:

> out_dir = 'out'
eval_interval = 2000
log_interval = 1
eval_iters = 200
eval_only = False # if True, script exits right after the first eval
always_save_checkpoint = True # if True, always save a checkpoint after each eval
init_from = 'scratch' # 'scratch' or 'resume' or 'gpt2*'
wandb_log = False # disabled by default
wandb_project = 'owt'
wandb_run_name = 'gpt2' # 'run' + str(time.time())
dataset = 'openwebtext'
gradient_accumulation_steps = 5 # used to simulate larger batch sizes
batch_size = 12 # if gradient_accumulation_steps > 1, this is the micro-batch size
block_size = 1024
n_layer = 12
n_head = 12
n_embd = 768
dropout = 0.0 # for pretraining 0 is good, for finetuning try 0.1+
bias = False # do we use bias inside LayerNorm and Linear layers?
learning_rate = 6e-4 # max learning rate
max_iters = 600000 # total number of training iterations
weight_decay = 1e-1
beta1 = 0.9
beta2 = 0.95
grad_clip = 1.0 # clip gradients at this value, or disable if == 0.0
decay_lr = True # whether to decay the learning rate
warmup_iters = 2000 # how many steps to warm up for
lr_decay_iters = 600000 # should be ~= max_iters per Chinchilla
min_lr = 6e-5 # minimum learning rate, should be ~= learning_rate/10 per Chinchilla

Here is the training summary along with logs on **1x 4090 GPU**:

- **~5 sec per iteration**

> iiter 0: loss 10.9774, time 24068.01ms, mfu -100.00%
iter 1: loss 10.9936, time 4675.04ms, mfu -100.00%
iter 2: loss 10.9437, time 4673.63ms, mfu -100.00%
iter 3: loss 10.9439, time 4676.30ms, mfu -100.00%
iter 4: loss 10.8917, time 4676.06ms, mfu -100.00%
iter 5: loss 10.8188, time 4675.66ms, mfu 28.80%
iter 6: loss 10.7433, time 4728.85ms, mfu 28.77%
iter 7: loss 10.6411, time 4696.99ms, mfu 28.76%
iter 8: loss 10.5707, time 4679.96ms, mfu 28.76%
iter 9: loss 10.4496, time 4687.36ms, mfu 28.76%
iter 10: loss 10.4046, time 4711.22ms, mfu 28.74%
iter 11: loss 10.3610, time 4667.22ms, mfu 28.75%
iter 12: loss 10.2007, time 4790.16ms, mfu 28.69%
iter 13: loss 10.1336, time 4751.91ms, mfu 28.65%
iter 14: loss 10.0921, time 4780.29ms, mfu 28.60%
iter 15: loss 10.0034, time 4800.29ms, mfu 28.55%
iter 16: loss 9.9804, time 4800.27ms, mfu 28.50%
iter 17: loss 9.9206, time 4750.19ms, mfu 28.48%
iter 18: loss 9.9195, time 4708.96ms, mfu 28.49%
iter 19: loss 9.8309, time 4717.19ms, mfu 28.50%
iter 20: loss 9.8023, time 4761.68ms, mfu 28.48%
iter 21: loss 9.7297, time 4846.96ms, mfu 28.41%
iter 22: loss 9.6467, time 4767.08ms, mfu 28.39%
iter 23: loss 9.6580, time 4802.63ms, mfu 28.36%
iter 24: loss 9.7079, time 4673.70ms, mfu 28.40%
iter 25: loss 9.7576, time 4653.34ms, mfu 28.46%
iter 26: loss 9.5448, time 4723.24ms, mfu 28.46%
iter 27: loss 9.5650, time 4745.47ms, mfu 28.45%
iter 28: loss 9.5764, time 4755.54ms, mfu 28.44%
iter 29: loss 9.5327, time 4650.28ms, mfu 28.49%
iter 30: loss 9.4981, time 4648.54ms, mfu 28.54%
iter 31: loss 9.4870, time 4650.06ms, mfu 28.58%
iter 32: loss 9.5341, time 4751.89ms, mfu 28.56%
iter 33: loss 9.3428, time 4789.42ms, mfu 28.51%
iter 34: loss 9.4235, time 4772.11ms, mfu 28.48%
iter 35: loss 9.4571, time 4766.35ms, mfu 28.46%
iter 36: loss 9.3941, time 4782.24ms, mfu 28.43%
iter 37: loss 9.3667, time 4674.66ms, mfu 28.47%
iter 38: loss 9.2519, time 4705.24ms, mfu 28.48%
iter 39: loss 9.2476, time 4756.26ms, mfu 28.47%
iter 40: loss 9.2942, time 4707.03ms, mfu 28.48%
iter 41: loss 9.2504, time 4683.16ms, mfu 28.51%
iter 42: loss 9.1140, time 4750.65ms, mfu 28.49%
iter 43: loss 9.2805, time 4760.89ms, mfu 28.47%
iter 44: loss 9.0174, time 4729.67ms, mfu 28.47%
iter 45: loss 9.3653, time 4754.12ms, mfu 28.46%
iter 46: loss 9.1722, time 4773.88ms, mfu 28.43%
iter 47: loss 9.0864, time 4728.74ms, mfu 28.44%
iter 48: loss 8.9944, time 4738.04ms, mfu 28.43%
iter 49: loss 8.9316, time 4768.75ms, mfu 28.41%
iter 50: loss 9.1234, time 4762.28ms, mfu 28.40%

I am running with latest master pulled today (02/24) that contains fixes for single-GPU training: https://github.com/karpathy/nanoGPT/pull/145

So for `max_iters = 600000` , total training time would be 600000 x 5 / (3600 x 24) = ~35 days while it took ~4 days using PyTorch Distributed Data Parallel (DDP) on 8x A100 40GB node. It seems like linear with respect to number of GPUs. I feel like even with 1x 4090, training time per iteration should have been much lower compared to 5 sec/iteration currently. If others can also share their training performance, it would help everyone to see the gains on 4090 GPU and make required fixes.

@otaviogood could you post your training time on 1x 4090 setup? I guess there are others who are using 4090 these days and this thread can serve as a good reference for others who don't have distributed training setup. TIA!

**Comment by vgoklani** [comment]
Hey @akjindal53244 I see that you're running CUDA 12 (via nvidia-smi). Did you build pytorch from source? What version of pytorch are you using?

**Comment by akjindal53244** [comment]
Hi @vgoklani , I am running CUDA 11.8 (nvidia-smi shows maximum supported CUDA version rather than actual version being used). I installed pytorch 2.0 via conda with `python 3.9`

![image](https://user-images.githubusercontent.com/5215386/221282523-b21fe058-9bd8-410b-9480-85949a42ae91.png)

torch version shows cuda 11.8

**Comment by rsliu94** [comment]
I think it is because this PR: https://github.com/karpathy/nanoGPT/commit/ae3a8d5fdd3ddb8b13fab182723476523961e3ab
The committer here multiplies `gradient_accumulation_steps `, so it will decrease your training speed on shakespeare_char by 40 times.

**Comment by akjindal53244** [comment]
Hi @rsliu94 , thank you for pointing this out. I was able to train `shakespeare_char` in 3 minutes on 1x 4090 GPU after removing the changes from commit.

**Comment by otaviogood** [comment]
I made a PR that fixes some of my errors from last time. Please LMK if there are other errors.
